fix(core): Keep identities when evaluate=False

oscarbenjamin opened the pull request:
Fixes #19337
Closes #13188

Changes Add/Mul to keep identities when evaluate=False.

Before:
```python
>>> Add(0, 1, evaluate=False)
1
```
After:
```python
>>> Add(0, 1, evaluate=False)
0 + 1
```
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

This is mentioned in many issues...

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
     * When creating an Add or Mul with evaluate=False identities (0 or 1) are no longer removed.
* printing
     * Printing unevaluated Muls with multiple Number factors will show all factors strictly in order. Identities will also be shown explicitly.
<!-- END RELEASE NOTES -->

mcpl-sympy commented:
Perhaps related to #18873

oscarbenjamin commented:
This is ready for review

mcpl-sympy commented:
I think this closes #13188

jksuom commented:
This looks good to me. The result of the example in https://github.com/sympy/sympy/pull/13188#issuecomment-324751681 is the same as there but that seems acceptable.

oscarbenjamin commented:
There is a problem with printing:
```julia
In [6]: with evaluate(False): 
   ...:     a = (S(1)/3) 
   ...:     b = 1*x 
   ...:                                                                                                                           

In [7]: a                                                                                                                         
Out[7]: 
1
─
3

In [8]: b                                                                                                                         
Out[8]: 1⋅x

In [9]: type(a)                                                                                                                   
Out[9]: sympy.core.mul.Mul

In [10]: a.args                                                                                                                   
Out[10]: 
⎛   1⎞
⎜1, ─⎟
⎝   3⎠

In [11]: srepr(a)                                                                                                                 
Out[11]: 'Mul(Integer(1), Pow(Integer(3), Integer(-1)))'
```
Perhaps it is also worth changing `One.__div__`...

smichr commented:
> There is a problem with printing:

If 1 is still in the args, this is not only a printing problem, right? `a` should be `Pow` with args `(3, -1)`.

oscarbenjamin commented:
> > There is a problem with printing:
> 
> If 1 is still in the args, this is not only a printing problem, right? `a` should be `Pow` with args `(3, -1)`.

I've fixed the printing problem which means that you can now "see" the other problem:
```julia
In [1]: with evaluate(False): 
   ...:     a = S(1)/5 
   ...:                                                                                                                           

In [2]: a                                                                                                                         
Out[2]: 
  1
1⋅─
  5

In [3]: str(a)                                                                                                                    
Out[3]: '1*1/5'
```

oscarbenjamin commented:
Just need to add a few tests and then this is done...

oscarbenjamin commented:
Looks like this leads to infinite recursion: `pretty(Mul(1.0, y))`

oscarbenjamin commented:
Another printing problem with `str(Mul(1, 1, evaluate=False))                                                                                           `

smichr commented on sympy/printing/tests/test_str.py:
```diff
@@ -224,6 +227,14 @@ def test_Mul():
     assert str(-2*x/3) == '-2*x/3'
     assert str(-1.0*x) == '-1.0*x'
     assert str(1.0*x) == '1.0*x'
+    assert str(Mul(0, 1, evaluate=False)) == '1*0'
```
Why is the order being reversed? This is different from the Add case. Does `Mul(1, 0, evaluate=0)` print the same?

oscarbenjamin replied:
I don't normally mess around with the printing code so I'm not entirely sure of the reasons for this but Mul does not print its args in order in general:
https://github.com/sympy/sympy/blob/aa9c901819d9390d5ed13a269bfc32569ea749b6/sympy/printing/str.py#L290-L294
This means that (on master):
```julia
In [1]: m = Mul(y, x, 2, 1, evaluate=False)                                                                                       

In [2]: m.args                                                                                                                    
Out[2]: (y, x, 2)

In [3]: str(m)                                                                                                                    
Out[3]: '2*x*y'

In [4]: m                                                                                                                         
Out[4]: 2⋅x⋅y
```
This PR gives
```julia
In [2]: m.args                                                                                                                    
Out[2]: (y, x, 2, 1)

In [3]: str(m)                                                                                                                    
Out[3]: '1*2*x*y'

In [4]: m                                                                                                                         
Out[4]: 1⋅2⋅x⋅y
```
Various manipulations are done that combine rationals e.g. (on master and here):
```julia
In [7]: str(Mul(S(2)/3, S(3)/4, evaluate=False))                                                                                  
Out[7]: '2*3/(3*4)'

In [8]: str(Mul(2, S(1)/3, S(3)/4, evaluate=False))                                                                               
Out[8]: '3*2/(3*4)'
```
The changes to the printing code here are just to stop identities being absorbed into other rationals so that you can at least see that you have a funny kind of object. This previously printed as 1 which was confusing while debugging:
```julia
In [9]: Mul(1, 1, evaluate=False)                                                                                                 
Out[9]: 1⋅1

In [10]: Mul(1, 1, evaluate=False).args                                                                                           
Out[10]: (1, 1)
```

smichr commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1786,6 +1790,15 @@ def _print_Mul(self, product):
         else:
             args = list(product.args)
 
+        # Always show the ones out front of a product:
```
Maybe it should be 0s then 1s then others.

oscarbenjamin replied:
I'd rather have the args displayed in their actual order but that's not what currently happens...

smichr replied:
Me too. Can we carry a "dirty" label on an expression if it is created with `evaluate=False` so we know (quickly) how it was created? That way the printer could skip the ordering in that case.

oscarbenjamin replied:
I guess we know if a Mul is unevaluated when it has more than one `Number` in the args or if the `Number` is `S.One` or is not in `args[0]` so:
```
def is_dirty(mul):
    args = mul.args
    if args[0] is S.One or any(isinstance(arg, Number) for arg in args[1:]):
        return True
```
Then in that case we can print the args multiplied precisely in order and not as a common fraction e.g.:
```julia
In [7]: Mul(1, S(1)/2, x, evaluate=False)                                                                                         
Out[7]:
  1
1⋅─⋅x
  2

In [14]: str(_)                                                                                                                   
Out[14]: '1*(1/2)*x'
```
That would also improve cases like:
```julia
In [21]: Mul(S(2)/3, S(7)/5, evaluate=False)                                                                                      
Out[21]: 
2⋅7
───
3⋅5

In [22]: str(_)                                                                                                                   
Out[22]: '2*7/(3*5)'
```

That doesn't cover all cases of unevaluated mul e.g. 2-arg mul or powers with the same exponent.

smichr commented:
> which means that you can now "see" the other problem:

How should that be fixed. We do have a div operator, so seeing it treated like mutliplying by the inverse seems to be a bit contrary.

oscarbenjamin commented:
> > which means that you can now "see" the other problem:
> 
> How should that be fixed. We do have a div operator, so seeing it treated like mutliplying by the inverse seems to be a bit contrary.

That's already fixed:
```julia
In [11]: with evaluate(False): 
    ...:     a = 1/x 
    ...:                                                                                                                          

In [12]: a                                                                                                                        
Out[12]: 
1
─
x

In [13]: type(a)                                                                                                                  
Out[13]: sympy.core.power.Pow

In [14]: a.args                                                                                                                   
Out[14]: (x, -1)
```

oscarbenjamin commented:
The only outstanding issue in this PR is whether to also change it so that `fraction(..., exact=True)` does not remove identities. Currently it does and that affects the latex printing since it uses `fraction`:
```julia
In [5]: latex(Mul(1, 2, S(1)/3, x, y, evaluate=False))                                                                            
Out[5]: '\\frac{2 x y}{3}'
```

smichr commented:
> Currently it does

You could see what happens if you just have it return an `evaluate=False` numerator and denominator. It is smart enough to treat `5**-1` as a denominator term so maybe it doesn't introduce unnecessary 1s?

smichr commented:
The explicit dirty flag would help with https://stackoverflow.com/questions/14624511/prevent-sympy-from-rearranging-the-equation/14667919#14667919

smichr commented:
```diff
diff --git a/sympy/simplify/radsimp.py b/sympy/simplify/radsimp.py
index 439d80f..9f6ba9f 100644
--- a/sympy/simplify/radsimp.py
+++ b/sympy/simplify/radsimp.py
@@ -1055,14 +1055,6 @@ def fraction(expr, exact=False):
        >>> fraction(exp(-x), exact=True)
        (exp(-x), 1)
 
-       The `exact` flag will also keep any unevaluated Muls from
-       being evaluated:
-
-       >>> u = Mul(2, x + 1, evaluate=False)
-       >>> fraction(u)
-       (2*x + 2, 1)
-       >>> fraction(u, exact=True)
-       (2*(x  + 1), 1)
     """
     expr = sympify(expr)
 
@@ -1085,22 +1077,18 @@ def fraction(expr, exact=False):
                 numer.append(term)
             elif not exact and ex.is_Mul:
                 n, d = term.as_numer_denom()
-                numer.append(n)
+                if n != 1:
+                    numer.append(n)
                 denom.append(d)
             else:
                 numer.append(term)
-        elif term.is_Rational:
-            n, d = term.as_numer_denom()
-            numer.append(n)
-            denom.append(d)
+        elif term.is_Rational and not term.is_Integer:
+            if term.p != 1:
+                numer.append(term.p)
+            denom.append(term.q)
         else:
             numer.append(term)
-    if exact:
-        numer = [n for n in numer if n != 1]
-        denom = [d for d in denom if d != 1]
-        return Mul(*numer, evaluate=False), Mul(*denom, evaluate=False)
-    else:
-        return Mul(*numer), Mul(*denom)
+    return Mul(*numer, evaluate=False), Mul(*denom, evaluate=False)
 
 
 def numer(expr):
diff --git a/sympy/simplify/tests/test_radsimp.py b/sympy/simplify/tests/test_radsimp.py
index 769ddcf..de68c8b 100644
--- a/sympy/simplify/tests/test_radsimp.py
+++ b/sympy/simplify/tests/test_radsimp.py
@@ -51,10 +51,10 @@ def test_radsimp():
     assert radsimp(1/(r2*a + r2*b + r3 + r7)) == (
         (sqrt(2)*a*(a + b)**2 - 5*sqrt(2)*a + sqrt(42)*a + sqrt(2)*b*(a +
         b)**2 - 5*sqrt(2)*b + sqrt(42)*b - sqrt(7)*(a + b)**2 - sqrt(3)*(a +
-        b)**2 - 2*sqrt(3) + 2*sqrt(7))/(2*a**4 + 8*a**3*b + 12*a**2*b**2 -
-        20*a**2 + 8*a*b**3 - 40*a*b + 2*b**4 - 20*b**2 + 8))
+        b)**2 - 2*sqrt(3) + 2*sqrt(7))/(a**4 + 4*a**3*b + 6*a**2*b**2 -
+        10*a**2 + 4*a*b**3 - 20*a*b + b**4 - 10*b**2 + 4)/2)
     assert radsimp(1/(r2*a + r2*b + r2*c + r2*d)) == \
-        sqrt(2)/(2*a + 2*b + 2*c + 2*d)
+        sqrt(2)/(a + b + c + d)/2
     assert radsimp(1/(1 + r2*a + r2*b + r2*c + r2*d)) == (
         (sqrt(2)*a + sqrt(2)*b + sqrt(2)*c + sqrt(2)*d - 1)/(2*a**2 + 4*a*b +
         4*a*c + 4*a*d + 2*b**2 + 4*b*c + 4*b*d + 2*c**2 + 4*c*d + 2*d**2 - 1))
```

smichr commented:
The diff gives
```python
>>> fraction(Mul(1,1,S.Half,evaluate=False))
(1, 2)
>>> fraction(Mul(1,1,S.Half,evaluate=False), exact=True)
(1*1, 2)
```

oscarbenjamin commented:
Thanks @smichr. I've pushed that and also changes to the latex printer.

oscarbenjamin commented:
I think this is ready.

smichr commented:
> I think this is ready.

Perhaps a squash to one commit?

oscarbenjamin committed:
fix(core): Keep identities when evaluate=False

Changes Add/Mul to keep identities when evaluate=False.

Before:
>>> Add(0, 1, evaluate=False)
1

After:
>>> Add(0, 1, evaluate=False)
0 + 1

A number of parts of the codebase are updated to be able to handle
identities in unevaluated Mul/Pow.

oscarbenjamin committed:
feat(printing): Print all args in unevaluated Muls

Change str, pretty and latex printers when handling unevaluated Muls. If
a Mul has two or more Numbers in its args or if the first arg is not a
Number then it must be an unevaluated Mul. In this case we print the
result strictly in order of args with no attempt to group numerators and
denominators together in a fraction so that the internal unevaluated
representation is shown clearly.

oscarbenjamin commented:
I've squashed it to two commits. One is for the changes to core and the other is the changes to printing.

smichr commented:
Looks good.
